Adding statistics backend for webhook

bidesh opened the pull request:
Currently, the httpNotification only works for notifying failures. To be able to have the same kind of statistics as cassandra has but not in cassandra, we have added a webhook integration as discussed in #564 with @andygrunwald. 
The JobStats is too coupled with Cassndra at the moment. I think seperating cassandra from JobStats is a good idea to be discussed. But this PR basically adds a publisher to push the data similar to Cassandra. Along with that, it also sends metrics in a json format. 
How data is sent:

``` json
{
    "attempt": "0",
    "histogram": {
        "75thPercentile": 14155.0,
        "95thPercentile": 16132.0,
        "98thPercentile": 16132.0,
        "99thPercentile": 16132.0,
        "count": 13,
        "mean": 13168.0,
        "median": 13168.0
    },
    "job_name": "RecurringSleep",
    "job_owner": "bidesh.thapaliya@trivago.com",
    "job_schedule": "R/2015-11-26T13:57:44.000Z/PT20S",
    "slave_id": "caa576d0-a3c8-494c-aad0-d2101ce6bc33-S1",
    "taskId": "ct:1448546204000:0:RecurringSleep:",
    "task_state": "TASK_RUNNING",
    "ts": "Thu Nov 26 13:56:44 UTC 2015"
}
```

This is the first approach. We would be very happy to get your feedback on this PR on your general interest in getting this feature into chronos.

brndnmtthws commented on src/main/scala/org/apache/mesos/chronos/scheduler/config/StatsPublisherConfiguration.scala:
```diff
@@ -0,0 +1,17 @@
+import org.rogach.scallop.ScallopConf
+
+/**
+  * Created by bthapaliya on 24/11/15.
```
Please remove.

brndnmtthws commented on src/main/scala/org/apache/mesos/chronos/scheduler/jobs/stats/JobStats.scala:
```diff
@@ -1,15 +1,19 @@
 package org.apache.mesos.chronos.scheduler.jobs.stats
 
+import com.fasterxml.jackson.core.JsonFactory
+
```
Follow the convention here please.

bidesh commented:
Hi @brndnmtthws,
Thanks for the feedback. I addressed your comment in the commit. But after that the test is failing and I can't figure out why my changes are affecting these tests. These changes were mainly removing the comments and repositioning the import statement. My guess was that something changed in the master that doesn't work properly with this PR.  I did the build locally by:

```
git fetch origin pull/593/head:prrequest593
git merge prrequest593
```

I ran the tests on `prrequest593` branch and on `master` branch after merging `prrequest593`. Both  cases seem to pass the tests properly. Would you have any idea on this? If not, I can investigate this further.

bidesh commented:
Hello @florianleibert @andyk @brndnmtthws,
Would it be possible to get a feedback on this? Do you think it would be possible to include such a feature in 2.5.0-RC1? 

Thank you.

Califax commented on src/main/scala/org/apache/mesos/chronos/scheduler/config/StatsPublisherConfiguration.scala:
```diff
@@ -0,0 +1,14 @@
+
+  lazy val webhookPort = opt[Int]("stats_webhook_port",
+    descr = "Port where the stats are pushed",
+    default = Some(8080))
```
Small, but the default specified is the default marathon port I believe. Given marathon is often used to run Chronos, might be a good idea to choose a different port.

bidesh replied:
Will do.

Califax commented on src/main/scala/org/apache/mesos/chronos/scheduler/jobs/stats/JobStats.scala:
```diff
@@ -562,4 +585,99 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
+    }
+    jobSchedule match {
+      case Some(js: String) => generator.writeStringField("job_schedule", js)
+      case _ =>
```
Could you add a test to show this works as expected. A lot of potential string typos.

bidesh replied:
Will make the changes and add tests.

Califax commented on src/main/scala/org/apache/mesos/chronos/scheduler/jobs/stats/JobStats.scala:
```diff
@@ -562,4 +585,99 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
+                                     timestamp: Option[java.util.Date],
+                                     jobName: Option[String],
+                                     jobOwner: Option[String],
+                                     jobSchedule: Option[String],
```
Is this the current schedule of the job?

bidesh replied:
Yes.

Califax commented on src/main/scala/org/apache/mesos/chronos/scheduler/jobs/stats/JobStats.scala:
```diff
@@ -562,4 +585,99 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
+                                     slaveId: Option[String],
+                                     message: Option[String],
+                                     attempt: Option[Integer],
+                                     isFailure: Option[Boolean]) = {
```
Can we add if the job is asychrononous or not?

bidesh replied:
We should be able to.

Califax commented on src/main/scala/org/apache/mesos/chronos/scheduler/config/StatsPublisherConfiguration.scala:
```diff
@@ -0,0 +1,14 @@
+
+trait StatsPublisherConfiguration extends ScallopConf{
+  lazy val webhookUrl = opt[String]("stats_webhook_url",
+    descr = "Url to which tasks statistics are pushed",
```
If for a store requiring authentication such as S3 where will the credentials be read from?

bidesh replied:
This is a good point. In that case, I have to extend the configuration for credentials. But would it make sense to include that in this PR? Extending it for all the possible stores would be diffucult. The initial idea was that we send the data to the hook and this hook has the logic to do whatever it wants. That would be a nice seperation from the chronos as well. I do agree, however, that there should be a way to provide credential to push to the hook.

Califax commented on src/main/scala/org/apache/mesos/chronos/scheduler/jobs/stats/JobStats.scala:
```diff
@@ -562,4 +585,99 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
+      case Some(a: Integer) => generator.writeStringField("attempt", a.toString())
+      case _ =>
+    }
+    isFailure match {
```
This does match the cassandra format, but it seems odd we use the negation isFailure over getSuccess. Out of scope for PR, just a thought.

Califax commented on src/main/scala/org/apache/mesos/chronos/scheduler/jobs/stats/JobStats.scala:
```diff
@@ -562,4 +585,99 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
+                                     jobName: Option[String],
+                                     jobOwner: Option[String],
+                                     jobSchedule: Option[String],
+                                     jobParents: Option[java.util.Set[String]],
```
I would like to dependent / children jobs of the current job. Option: We can also indicate whether it would kick off a child (looking at invocationmap) if the job was successful. Would be great to see for logging purposes / detecting any failures in dependency management.

bidesh replied:
I will get back to you on this one.

brndnmtthws commented:
I'm closing this PR for now. If you'd like to rebase against master and reopen, feel free.
